### PR TITLE
Redefine SendCodecs and ReceiveCodecs

### PR DESCRIPTION
**Pull request by alvestrand**
to contain all supported codecs, revealing only those that are "enabled" at any given time.

Fixes: #2925

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webrtc-pc/pull/2935.html" title="Last updated on Feb 23, 2024, 2:15 PM UTC (c2d1b55)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webrtc-pc/2935/3d6e620...c2d1b55.html" title="Last updated on Feb 23, 2024, 2:15 PM UTC (c2d1b55)">Diff</a>

### CODE REVIEW

**Comment by alvestrand** [comment]
Not sure "editorial" is correct here. It does document the existing browser behavior that implemented but not announced codecs are revealed when they are specifically asked for. But it unblocks the check.

File: webrtc.html [hunk 1]
```diff
@@ -2573,11 +2577,12 @@ <h4>
                               </li>
                               <li>
                                 <p>
-                                  Set
-                                  <var>transceiver</var>.{{RTCRtpTransceiver/[[Receiver]]}}.{{RTCRtpReceiver/[[ReceiveCodecs]]}}
-                                  to the codecs that <var>description</var>
-                                  negotiates for receiving and which the user
-                                  agent is currently prepared to receive.
+                                  For each of the codecs that <var>description</var> negotiates for receiving:
+                                      <ol>
```
Comment (alvestrand) [inline webrtc.html:2581]: execute the following steps

File: webrtc.html [hunk 2]
```diff
@@ -2590,11 +2595,12 @@ <h4>
                                 <ol>
                                   <li>
                                     <p>
-                                      Set
-                                      <var>transceiver</var>.{{RTCRtpTransceiver/[[Sender]]}}.{{RTCRtpSender/[[SendCodecs]]}}
-                                      to the codecs that <var>description</var>
-                                      negotiates for sending and which the user
-                                      agent is currently capable of sending.
+                                      For each of the codecs that <var>description</var> negotiates for sending:
+                                      <ol>
```
Comment (alvestrand) [inline webrtc.html:2599]: execute the following steps

File: webrtc.html [hunk 3]
```diff
@@ -2246,11 +2246,13 @@ <h4>
...
-                                  Set
-                                  <var>transceiver</var>.{{RTCRtpTransceiver/[[Receiver]]}}.{{RTCRtpReceiver/[[ReceiveCodecs]]}}
-                                  to the codecs that <var>description</var>
-                                  negotiates for receiving and which the user
-                                  agent is currently prepared to receive.
+                                  For each of the codecs that <var>description</var> negotiates for receiving:
+                                      <ol>
+                                        <li>Locate the matching codec description in <var>transceiver</var>.{{RTCRtpTransceiver/[[Receiver]]}}.{{RTCRtpReceiver/[[ReceiveCodecs]]}}.</li>
+                                        <li>If the matching codec description is not found, abort these steps.</li>
+                                        <li>Set the "enabled" flag in the matching codec description to "true".</li>
```
[the remark this answers is not in the document]
Comment (alvestrand) [reply webrtc.html:2253]: How? Wouldn't that start at the defintion of the codec info?
(never used "tuple" in WebIDL before)

**Comment by alvestrand** [comment]
WG discussion on Feb 20, 2024: Looks like an improvement.

File: webrtc.html [hunk 4]
```diff
@@ -8851,8 +8854,10 @@ <h3>
           <li>
             <p>
               Let <var>sender</var> have a <dfn class="export" data-dfn-for="RTCRtpSender">[[\SendCodecs]]</dfn> internal
-              slot, representing a list of {{RTCRtpCodecParameters}}
-              dictionaries, and initialized to an empty list.
+              slot, representing a list where each element contains an {{RTCRtpCodecParameters}}
```
Comment (alvestrand) [inline webrtc.html:8857]:
```suggestion
              slot, representing a list of tuples containing an {{RTCRtpCodecParameters}}
```

**Review by jan-ivar — APPROVED** [review]

**Comment by alvestrand** [comment]
From @jan-ivar : link definition of "enabled" to a dfn so that what it is is defined in one place.

File: webrtc.html [hunk 5]
```diff
@@ -8851,8 +8854,10 @@ <h3>
           <li>
             <p>
               Let <var>sender</var> have a <dfn class="export" data-dfn-for="RTCRtpSender">[[\SendCodecs]]</dfn> internal
-              slot, representing a list of {{RTCRtpCodecParameters}}
-              dictionaries, and initialized to an empty list.
+              slot, representing a list of tuples, each containing an {{RTCRtpCodecParameters}}
```
Comment (alvestrand) [inline webrtc.html:8857]:
```suggestion
              slot, representing a list of [=tuple=]s, each containing an {{RTCRtpCodecParameters}}
```

**Comment by alvestrand** [comment]
Hoisting the @dontcallmedom flag. Can't figure out how to run respec locally to get the error message that listAmendments is complaining about.

**Comment by dontcallmedom** [comment]
FYI: what was missing was the step of 1 of the [amendment markup process](https://github.com/w3c/webrtc-pc/blob/main/amendments.md) - adding the id of the changed section to `base-rec.html`.

**Comment by alvestrand** [comment]
Thank you @dontcallmedom ! (and for reminding me of the documentation link!!)

**Comment by dontcallmedom** [comment]
are there tests associated with this pull request? I would document them in #2910 (and if not, we should track the need to create them)

**Pull request merged**